File: 21/MergeTwoSortedLists.c
struct ListNode{
    int val;
    struct ListNode* next;
};

#define Null 0

/**
 * Note: specify Null if not specified.
 * Note: Null is not always 0 in IA64
 * Note: '->' is used for a struct pointer to get element inside.
 * Note: new a struct should use (Ptr*)malloc(sizeof(Ptr))
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
    if (l1 == Null) {
        return l2;
    } else if (l2 == Null) {
        return l1;
    }
    struct ListNode *r = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *p = r;
    while (l1 != Null || l2 != Null) {
        if (l1 != Null && l2 != Null) {
            if (l1->val < l2->val) {
                p->next = l1;
                l1 = l1->next;
            } else {
                p->next = l2;
                l2 = l2->next;
            }
        } else if (l2 != Null) {
            p->next = l2;
            l2 = l2->next;
        } else {
            p->next = l1;
            l1 = l1->next;
        }
        p = p->next;
    }
    return r->next;
}

int main() {
    struct ListNode* l1 = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode* l2 = (struct ListNode*)malloc(sizeof(struct ListNode));
    mergeTwoLists(l1, l2);
}

File: 21/MergeTwoSortedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * Note: mind the margins
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    if (!l1) {
        return l2;
    } else if (!l2) {
        return l1;
    }
    var des = false;
    if (l1.next) {
        des = l1.val > l1.next.val;
    } else if (l2.next) {
        des = l2.val > l2.next.val;
    }
    var r = new ListNode();
    var p = r;
    while (l1 || l2) {
        if (l1 && l2) {
            if ((l1.val > l2.val && des) || (l1.val < l2.val && !des)) {
                p.next = l1;
                l1 = l1.next;
            } else {
                p.next = l2;
                l2 = l2.next;
            }
        } else if (l2) {
            p.next = l2;
            l2 = l2.next;
        } else {
            p.next = l1;
            l1 = l1.next;
        }
        p = p.next;
    }
    return r.next;
};